File: tp1/source/ucs.py
from graph_search import GraphSearch

from priority_frontier import PriorityFrontier


class UCS(GraphSearch):

    def _init_frontier(self, dimensions, node):
        self.frontier = PriorityFrontier(dimensions)
        self.frontier.insert((node.path_cost, node))

    def _update_frontier(self, child):
        if child.state in self.frontier:
            self.frontier.replace((child.path_cost, child))
        elif child.state not in self.explored:
            self.frontier.insert((child.path_cost, child))

File: tp1/source/ucs.cpp
#include "ucs.h"


void UCS::init_frontier(int height, int width){
    this->frontier = std::priority_queue<Node>;
    this->init_frontier_hash(height, width);
}

void UCS::replace_insert_frontier(const Node& n){
    for(int i = 0;)
    for i, node in enumerate(self.frontier):
        if node.state == item.state:
            if node.path_cost > item.path_cost:
                self.frontier[i] = item
                heapq.heapify(self.frontier)
            break
}

void UCS::update_frontier(const Node& child){
    if(this->in_frontier(child)){
        // if child.STATE is in frontier with higher PATH-COST
        // then replace frontier node with child
        self._replace_insert_frontier(child);
    }
    else if(!this->in_explored(child)){
        this->insert_frontier(child);
    }
}
